Implementation/4 implement person service

stishchenko opened the pull request:
PersonServiceImpl created to implement IsPersonService. All methods use Stream api.

rdovgan committed:
Added IsPersonService for implementation

rdovgan committed:
Added test classes for PersonService

Tishchenko666 committed:
#4 PersonServiceImpl created to implement IsPersonService.
All methods use Stream api.
Method collectPersonsWithNameStartsWith() filter null-objects and objects without name, then leave objects with name starts with prefix.
Method collectPersonsByAge() filter null-objects and objects with incorrect age, then sort and group objects by age.
Method calculateAverageAge() filter null-objects and objects with incorrect age, then map to int and get average from summary statistics.
Method sumAndCountAge() filter null-objects and objects with incorrect age, then map to int and get summary statistics.

rdovgan reviewed (changes requested):
Need to wrap with `{}` even single line.
E.g.:
```
if (CollectionUtils.isEmpty(personList)) {
			return Collections.emptyList();
}
```

rdovgan commented on src/main/java/team/asd/service/PersonServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+	@Override
+	@NonNull
+	public List<IsPerson> collectPersonsWithNameStartsWith(List<IsPerson> personList, String prefix) {
+		if (personList == null || personList.isEmpty())
```
Use `CollectionUtils.isEmpty()` method

rdovgan commented on src/main/java/team/asd/service/PersonServiceImpl.java:
```diff
@@ -0,0 +1,66 @@
+	public List<IsPerson> collectPersonsWithNameStartsWith(List<IsPerson> personList, String prefix) {
+		if (personList == null || personList.isEmpty())
+			return new ArrayList<>();
+		if (prefix == null || prefix.equals(""))
```
Use `StringUtils.isEmpty()` method

Tishchenko666 committed:
Merge branch 'master' into implementation/4_Implement_PersonService

Tishchenko666 committed:
PersonServiceImpl was refactored. If clauses with single line get {}.
Checking null and empty clauses in method collectPersonsWithNameStartsWith() was changed on CollectionUtils.isEmpty() and new list creation - on Collections.emptyList().
In method collectPersonsByAge() new map creation was changed on Collections.emptyMap().

NckGrs commented on src/main/java/team/asd/service/PersonServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+			return new IntSummaryStatistics();
+		}
+		return personList.stream()
+				.filter(p -> p != null && p.getAge() != null && p.getAge() >= 0)
```
`personList.stream().filter(p -> p != null && p.getAge() != null && p.getAge() >= 0)`
This code is repeated three times, you can replace it with a separate method

NckGrs commented on src/main/java/team/asd/service/PersonServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+		if (CollectionUtils.isEmpty(personList)) {
+			return Collections.emptyList();
+		}
+		if (prefix == null || prefix.equals("")) {
```
`prefix == null || prefix.equals("")`
You can use the StringUtils.isEmpty() method instead

NckGrs commented on src/main/java/team/asd/service/PersonServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+			return personList;
+		}
+		return personList.stream()
+				.filter(p -> p != null && p.getName() != null && p.getName()
```
`p.getName() != null && p.getName().startsWith(prefix)`
You can use StringUtils.startsWith( ... ) or StringUtils.startsWithIgnoreCase( ... ) methods instead.

NckGrs commented on src/main/java/team/asd/service/PersonServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+		}
+		personList = personList.stream()
+				.filter(p -> p != null && p.getAge() != null && p.getAge() >= 0)
+				.collect(Collectors.toList());
```
I could be wrong, but most likely you don't need an intermediate collection of the stream into a list with further validation. Also you can use the Collectors.averagingInt(IsPerson::getAge) collector instead of 
`.mapToInt(IsPerson::getAge)
				.summaryStatistics()
				.getAverage();`

stishchenko replied:
I do this validation to be sure that after filtering by age it is not empty. If it is i have to return 0D instead of counting average. 
Thanks for collector. It looks better.

NckGrs replied:
https://www.geeksforgeeks.org/java-8-collectors-averagingint-in-java-with-examples/
For an empty collection, the method returns 0.0d

NckGrs commented on src/main/java/team/asd/service/PersonServiceImpl.java:
```diff
@@ -0,0 +1,72 @@
+		}
+		return personList.stream()
+				.filter(p -> p != null && p.getAge() != null && p.getAge() >= 0)
+				.sorted(Comparator.comparing(IsPerson::getAge))
```
There is nothing about sorting in the method description

stishchenko replied:
You are right. I have thought about sorting because of test, but it works without it too.

Tishchenko666 committed:
Replaced repeatable code that check age by separate method checkAge().

Tishchenko666 committed:
In method collectPersonsWithNameStartsWith() replace prefix checking by StringUtils.isEmpty().
Predicate filterPersonsByAge is created for filtering persons by age. It is used in second, third and fourth methods.

Tishchenko666 committed:
In method collectPersonsWithNameStartsWith() in filter name checking was replaced by StringUtils.
In method calculateAverageAge() map and SummaryStatistics was replaced by collector by age with counting the average.
